<a href="https://colab.research.google.com/github/Adeela-saleem/MachineLearning_Algorithms/blob/main/Supervised_Learning/Regression/RidgeRegularization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:

from sklearn.datasets import load_diabetes

data=load_diabetes()

In [3]:
print(data.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

:Number of Instances: 442

:Number of Attributes: First 10 columns are numeric predictive values

:Target: Column 11 is a quantitative measure of disease progression one year after baseline

:Attribute Information:
    - age     age in years
    - sex
    - bmi     body mass index
    - bp      average blood pressure
    - s1      tc, total serum cholesterol
    - s2      ldl, low-density lipoproteins
    - s3      hdl, high-density lipoproteins
    - s4      tch, total cholesterol / HDL
    - s5      ltg, possibly log of serum triglycerides level
    - s6      glu, blood sugar level

Note: Each of these 10 feature variables have bee

In [4]:
X=data.data

In [5]:
y=data.target

In [6]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=45)

In [7]:

from sklearn.linear_model import LinearRegression
L=LinearRegression()



In [8]:
L.fit(X_train,y_train)

LinearRegression()

In [9]:
print(L.coef_)
print(L.intercept_)

[  23.45465406 -247.42747406  492.1087518   329.35876431 -970.79723039
  573.54295519  182.42162368  255.92168168  794.21609282   89.32249214]
152.13623331746496


In [10]:
y_pred=L.predict(X_test)

In [11]:
from sklearn.metrics import r2_score,mean_squared_error

print("R2 score",r2_score(y_test,y_pred))
print("RMSE",np.sqrt(mean_squared_error(y_test,y_pred)))

R2 score 0.5188113124539249
RMSE 48.72713760953253


In [12]:

from sklearn.linear_model import Ridge
R=Ridge(alpha=100000)

In [13]:
R.fit(X_train,y_train)

Ridge(alpha=100000)

In [14]:
print(R.coef_)
print(R.intercept_)

[ 0.00260126  0.00057066  0.00776597  0.0060976   0.00233864  0.00184724
 -0.00513942  0.0052716   0.00734598  0.00528629]
151.83287930791352


In [15]:
y_pred1=R.predict(X_test)

In [16]:
print("R2 score",r2_score(y_test,y_pred1))
print("RMSE",np.sqrt(mean_squared_error(y_test,y_pred1)))

R2 score -0.00042490200441935855
RMSE 70.25956272917782


### **Ridge Regularization Behavior Analysis**

1. **Impact of Over-Regularization ($\alpha = 100,000$):**
   * Setting an excessively high penalty term ($\alpha$) forces all regression coefficients ($\boldsymbol{\beta}$) close to zero ($\boldsymbol{\beta} \approx \mathbf{0}$).
   * **Learned Coefficients:** Drops from large values (e.g., $-970.79$) to near-zero ($0.0023$).

2. **Performance Collapse (Underfitting):**
   * Baseline Linear Regression $R^2$: **~0.518**
   * Ridge Regression ($\alpha=100,000$) $R^2$: **~ -0.0004**
   * When $\alpha \to \infty$, the model simplifies to predicting only the target mean ($\hat{y} \approx \bar{y}$), leading to standard severe underfitting.

3. **Key Rule:**
   * $\alpha$ needs proper tuning (e.g., using `RidgeCV` or GridSearch over $10^{-3} \dots 10^2$) to balance bias-variance tradeoff without crushing model capacity.

In [17]:
from sklearn.linear_model import Ridge
from sklearn.metrics import r2_score

alphas = [0.0001, 0.001, 0.01, 0.1, 1, 10, 100, 1000]

for a in alphas:
    r = Ridge(alpha=a)
    r.fit(X_train, y_train)
    y_pred = r.predict(X_test)
    print(f"Alpha: {a:7} | R2 Score: {r2_score(y_test, y_pred):.5f}")

Alpha:  0.0001 | R2 Score: 0.51897
Alpha:   0.001 | R2 Score: 0.52014
Alpha:    0.01 | R2 Score: 0.52308
Alpha:     0.1 | R2 Score: 0.52084
Alpha:       1 | R2 Score: 0.45671
Alpha:      10 | R2 Score: 0.18325
Alpha:     100 | R2 Score: 0.02522
Alpha:    1000 | R2 Score: 0.00222


### **Optimal Alpha Selection Findings**

1. **Best Hyperparameter Value:**
   * $\alpha = 0.01$ yields the optimal test $R^2$ score of **~0.5231**, outperforming ordinary Linear Regression (**~0.5188**).

2. **Regularization Tradeoff:**
   * **Small Penalty ($\alpha \le 0.1$):** Properly controls variance without inflating bias.
   * **Large Penalty ($\alpha \ge 1$):** Forces severe underfitting, causing $R^2$ score to drop drastically towards 0.

 Practical Machine Learning WorkflowReal-world ML pipeline mein standard workflow yeh hota hai:Step 1: Pehle simple Linear Regression as a baseline chalayein.Step 2: Training score aur Testing score ko compare karein:Agar Train Score $\approx$ Test Score $\rightarrow$ Ridge ki zaroorat nahi hai.Agar Train Score $\gg$ Test Score (Overfitting) $\rightarrow$ Ridge/Lasso apply karein.Step 3: Automated cross-validation tool (jaise RidgeCV ya GridSearchCV) se optimal penalty value $\alpha$ dhoond kar model finalize karein.